lib/LaTeXML/Post/Manifest/Epub.pm

```diff
@@ -143,6 +167,16 @@ sub initialize {
   $$self{nav_map}       = $nav_map;
   return; }
 
+sub file_id {
+  my ($name) = @_;
+  # convert file name to valid NCName for use as id
+  # any invalid character, and -, are converted to -xN- where N is the hex codepoint
+  $name =~ s/([^A-Z_a-z\x{C0}-\x{D6}\x{D8}-\x{F6}\x{F8}-\x{2FF}\x{370}-\x{37D}\x{37F}-\x{1FFF}\x{200C}-\x{200D}\x{2070}-\x{218F}\x{2C00}-\x{2FEF}\x{3001}-\x{D7FF}\x{F900}-\x{FDCF}\x{FDF0}-\x{FFFD}\x{10000}-\x{EFFFF}.0-9\x{B7}\x{0300}-\x{036F}\x{203F}-\x{2040}])/-x${\(sprintf("%X", ord($1)))}-/g;
+  # ensure the starting char is valid and prevent collisions with the other id's
+  $name = 'file-' . $name;
+  return $name;
```

Comment:
I converted a large book from arXiv and took a look through the OPF manifest ids. Here are some curious entries:
```xml
<item id="file-images-x2F-GD-x2D-nosuccess-x2D-cont.png" href="images/GD-nosuccess-cont.png" media-type="image/png"/>
```

The href points to the correct file, but I wonder why the id got the `X2F` and `X2D` entries? They look like the hex codes for slash and dash. Well, the dash could have stayed as-is, and the X2F could have become the dash, keeping a slightly easier to read:
```xml
<item id="file-images-GD-nosuccess-cont.png" href="images/GD-nosuccess-cont.png" media-type="image/png"/>
```

which I think is closer to the state from the master branch.

Also, the leading `file-` prefix is an interesting choice, if it is to help avoid clashes, maybe a leading `opf-` instead is both more unique and slightly clearer? So my example may end up:

```xml
<item id="opf-images-GD-nosuccess-cont.png" href="images/GD-nosuccess-cont.png" media-type="image/png"/>
```

Reply:
> why the id got the `X2F` and `X2D` entries?

We would run a tiny risk of collision e.g. `math/0.svg` and `math-0.svg` get the same id (unlikely to happen, I know).

I can do some simplifications to get more readable entries (two lines of code): (1) encode `-` only when ambiguous (2) use `_` instead of `-` for the encoding (3) use the `URI::file` value to make Windows and Unix output the same id's [uh, (3) is a good idea regardless]. How does that sound @brucemiller ?

It should be noted that these id's are not visible because they are not used outside of the manifest (e.g. they are not anchors), so I'd prioritize avoiding collisions over readability.

> maybe a leading `opf-` instead is both more unique and slightly clearer

Actually, these id's are scoped to the OPF document, so `opf-` would be redundant. The other id's we contend with are `pub-id` (not a file), `nav` (...a file, actually). Here `file-` would just mean "actual file", or maybe "file generated before `Epub.pm`", does that make sense?

Reply:
I went for all (1)-(3), now the above example would become
```html
<item id="url-images_x2F_GD-nosuccess-cont.png" href="images/GD-nosuccess-cont.png" media-type="image/png"/>
```
and underscore would normally be preserved unless already in a sequence like `_x2F_`.

The code is slightly more complicated because of the lookahead in the regex.

Reply:
I like `url-` a lot more than `file-`, no reason why we can't both avoid conflicts and have a naming convention that everyone enjoys :) Of course, unclear if Bruce enjoys the x2F (compared to e.g. "slash", or double underscore or...), but it's up to him to finish the review here. I'd probably opt for the most human-friendly variant.

Generally happy with a merge here since it cleans things up a bit, can be tweaked as people stress-test it against larger ePub projects.

Reply:
OK, I finally took a look saw where the "ids" are coming in, and they're really only aliases to the `href` and purely internal to the manifest; barely ids at all, from my perspective :>   You could probably come up with something prettier, and even avoid conflicts, but with messier code... and really, who's ever going to see it except @dginev ? :>

So, yeah, it's ugly, but ok.